### PR TITLE
Bluetooth: Controller: Reset dup filtering on Periodic Re-sync 

### PR DESCRIPTION
**Pull request by cvinayak**
Reset duplicate filtering of Periodic Advertising Reports
when Periodic Advertising Sync is created again after a
terminate or sync lost.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/controller/hci/hci.c [hunk 1]
```diff
@@ -4876,18 +4897,54 @@ static void dup_ext_adv_mode_reset(struct dup_ext_adv_mode *dup_adv_mode)
...
 		dup_ext_adv_mode_reset(dup->adv_mode);
 	}
 }
+
+static void dup_periodic_adv_reset(uint8_t addr_type, const uint8_t *addr,
+				   uint8_t sid)
+{
+	int addr_idx;
+
+	for (addr_idx = 0; addr_idx < dup_count; addr_idx++) {
```
Comment (Thalley) [inline subsys/bluetooth/controller/hci/hci.c:4918]:
```suggestion
	for (int addr_idx = 0; addr_idx < dup_count; addr_idx++) {
```

or 

```suggestion
	for (int32_t addr_idx = 0; addr_idx < dup_count; addr_idx++) {
```

If you want to use the same type for `addr_idx` and `dup_count`.

File: subsys/bluetooth/controller/hci/hci.c [hunk 2]
```diff
@@ -4876,18 +4897,54 @@ static void dup_ext_adv_mode_reset(struct dup_ext_adv_mode *dup_adv_mode)
...
+		struct dup_ext_adv_mode *dup_mode;
+		struct dup_entry *dup;
+		int set_idx;
+
+		dup = &dup_filter[addr_idx];
+		if (memcmp(addr, &dup->addr.a.val[0],
```
Comment (Thalley) [inline subsys/bluetooth/controller/hci/hci.c:4924]:
```suggestion
		if (memcmp(addr, dup->addr.a.val,
```

File: subsys/bluetooth/controller/hci/hci.c [hunk 3]
```diff
@@ -4876,18 +4897,54 @@ static void dup_ext_adv_mode_reset(struct dup_ext_adv_mode *dup_adv_mode)
...
+
+static void dup_periodic_adv_reset(uint8_t addr_type, const uint8_t *addr,
+				   uint8_t sid)
+{
+	int addr_idx;
+
+	for (addr_idx = 0; addr_idx < dup_count; addr_idx++) {
+		struct dup_ext_adv_mode *dup_mode;
+		struct dup_entry *dup;
+		int set_idx;
```
Comment (Thalley) [inline subsys/bluetooth/controller/hci/hci.c:4921]:
```suggestion
		uint16_t set_idx;
```
If you want `set_idx` to be same type as `dup_mode->set_count` which it is compared against.

File: subsys/bluetooth/controller/hci/hci.c [hunk 4]
```diff
@@ -6246,31 +6314,39 @@ static void le_per_adv_sync_report(struct pdu_data *pdu_data,
...
-						sync->peer_id_addr_type,
-						sync->peer_id_addr,
-						DUP_EXT_ADV_MODE_PERIODIC,
-						adi, 0U);
-	}
+		uint8_t data_status;
+
+		data_status = (aux_ptr) ?
+			      BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL :
+			      BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE;
```
Comment (Thalley) [inline subsys/bluetooth/controller/hci/hci.c:6328]:
```suggestion
		const uint8_t data_status = (aux_ptr) ?
			      BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL :
			      BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE;
```

File: subsys/bluetooth/controller/hci/hci.c [hunk 5]
```diff
@@ -6246,31 +6314,39 @@ static void le_per_adv_sync_report(struct pdu_data *pdu_data,
...
+
+		drop = !ftr->sync_rx_enabled ||
+		       (sync->nodups && dup_found(PDU_ADV_TYPE_EXT_IND,
+						  sync->peer_id_addr_type,
+						  sync->peer_id_addr,
+						  DUP_EXT_ADV_MODE_PERIODIC,
+						  adi, data_status));
```
Comment (Thalley) [inline subsys/bluetooth/controller/hci/hci.c:6335]: Consider moving the result of `dup_found` to a separate variable for readability.

File: subsys/bluetooth/controller/hci/hci.c [hunk 6]
```diff
@@ -6113,10 +6181,10 @@ static void le_per_adv_sync_report(struct pdu_data *pdu_data,
 	uint8_t data_len_max;
 	uint8_t *acad = NULL;
 	uint8_t hdr_buf_len;
-	bool dup = false;
 	uint8_t hdr_len;
 	uint8_t *ptr;
 	int8_t rssi;
+	bool drop;
```
Comment (Thalley) [inline subsys/bluetooth/controller/hci/hci.c:6187]: It looks like `drop` is only used negated (i.e. as `!drop`). In such cases, I think it is better to revert the value, i.e. rename this to `no_drop` or `keep` as avoid the several negations below.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c [hunk 7]
```diff
@@ -616,6 +616,7 @@ static void isr_aux_setup(void *param)
 static int isr_rx(struct lll_sync *lll, uint8_t node_type, uint8_t crc_ok, uint8_t rssi_ready,
 		  enum sync_status status)
 {
+	uint8_t sched = 0U;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:619]:
```suggestion
	bool sched = false;
```

And update the following uses of this to true/false as well.

File: subsys/bluetooth/controller/ll_sw/ull_sync.c [hunk 8]
```diff
@@ -349,8 +350,24 @@ uint8_t ll_sync_terminate(uint16_t handle)
...
+	}
+
+	/* Reporting enabled/disabled */
+	sync->rx_enable = (enable & BT_HCI_LE_SET_PER_ADV_RECV_ENABLE_ENABLE) ?
+			  1U : 0U;
+
+#if defined(CONFIG_BT_CTLR_SYNC_PERIODIC_ADI_SUPPORT)
+	sync->nodups = (enable &
+			BT_HCI_LE_SET_PER_ADV_RECV_ENABLE_FILTER_DUPLICATE) ?
+		       1U : 0U;
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull_sync.c:367]: It's OK to go past 80 characters (and below 100) if it improves readability significantly, which I think this does:

```suggestion
	sync->nodups = (enable & BT_HCI_LE_SET_PER_ADV_RECV_ENABLE_FILTER_DUPLICATE) ?
		       1U : 0U;
```

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c (hunk 7 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:619]: Fixed.

File: subsys/bluetooth/controller/ll_sw/ull_sync.c (hunk 8 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_sync.c:367]: Fixed

**Review by cvinayak — COMMENTED** [review]
Some of the changes are fixed in #40832

File: subsys/bluetooth/controller/hci/hci.c [hunk 9]
```diff
@@ -4906,18 +4910,50 @@ static void dup_ext_adv_mode_reset(struct dup_ext_adv_mode *dup_adv_mode)
...
+static void dup_periodic_adv_reset(uint8_t addr_type, const uint8_t *addr,
+				   uint8_t sid)
+{
+	for (int32_t addr_idx = 0; addr_idx < dup_count; addr_idx++) {
+		struct dup_ext_adv_mode *dup_mode;
+		struct dup_entry *dup;
+
+		dup = &dup_filter[addr_idx];
+		if (memcmp(addr, dup->addr.a.val,
+			   sizeof(bt_addr_t)) ||
```
Comment (Thalley) [inline subsys/bluetooth/controller/hci/hci.c:4933]:
```suggestion
		if (memcmp(addr, dup->addr.a.val, sizeof(bt_addr_t)) ||
```

File: subsys/bluetooth/controller/hci/hci.c [hunk 10]
```diff
@@ -4959,6 +4995,13 @@ static inline bool is_dup_or_update(struct dup_entry *dup, uint8_t adv_type,
 			if (adv_set->adi.did != adi->did) {
 				/* report different DID */
 				adv_set->adi.did = adi->did;
+				/* set new data status */
+				if (data_status ==
+				    BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE) {
```
Comment (Thalley) [inline subsys/bluetooth/controller/hci/hci.c:5000]:
```suggestion
				if (data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE) {
```

File: subsys/bluetooth/controller/hci/hci.c [hunk 11]
```diff
@@ -4959,6 +4995,13 @@ static inline bool is_dup_or_update(struct dup_entry *dup, uint8_t adv_type,
...
+					adv_set->data_cmplt = 1U;
+				} else {
+					adv_set->data_cmplt = 0U;
+				}
```
Comment (Thalley) [inline subsys/bluetooth/controller/hci/hci.c:5004]:
```suggestion
				}

```

**Review by Thalley — APPROVED** [review]
Minor indention style comments, otherwise LGTM

**Review by carlescufi — APPROVED** [review]

**Pull request merged**